src/abstractdataframe/join.jl

```diff
@@ -91,6 +91,231 @@ _rename_cols(old_names::AbstractVector{Symbol},
+        cols[i+ncleft] = col
+    end
+
+    new_names = vcat(_rename_cols(_names(joiner.dfl), left_rename, joiner.left_on),
+                     _rename_cols(_names(dfr_noon), right_rename))
+    res = DataFrame(cols, new_names, makeunique=makeunique, copycols=false)
+
+    return res, nothing, nothing
+end
+
+@inline function find_next_range(x::AbstractArray, start::Int, start_value)
+    local stop_value
+    n = length(x)
+    stop = start + 1
+    while stop <= n
+        stop_value = x[stop]
+        if isequal(start_value, stop_value)
+            stop += 1
+        else
+            return stop, stop_value
+        end
+    end
+    return stop, start_value
+end
+
+function _innerjoin_sorted(left::AbstractArray, right::AbstractArray)
+    left_n = length(left)
+    right_n = length(right)
+
+    left_ixs = Int[]
+    right_ixs = Int[]
+
+    (left_n == 0 || right_n == 0) && return left_ixs, right_ixs
+
+    # lower bound assuming we get matches
+    sizehint!(left_ixs, min(left_n, right_n))
+    sizehint!(right_ixs, min(left_n, right_n))
+
+    left_cur = 1
+    left_val = left[left_cur]
+    left_new, left_tmp = find_next_range(left, left_cur, left_val)
+
+    right_cur = 1
+    right_val = right[right_cur]
+    right_new, right_tmp = find_next_range(right, right_cur, right_val)
+
+    while left_cur <= left_n && right_cur <= right_n
+        if isequal(left_val, right_val)
+            if left_new - left_cur == right_new - right_cur == 1
+                push!(left_ixs, left_cur)
+                push!(right_ixs, right_cur)
+            else
+                for (left_i, right_i) in Iterators.product(left_cur:left_new - 1,
+                                                           right_cur:right_new - 1)
+                    push!(left_ixs, left_i)
+                    push!(right_ixs, right_i)
+                end
+            end
+            left_cur, left_val = left_new, left_tmp
+            left_new, left_tmp = find_next_range(left, left_cur, left_val)
+            right_cur, right_val = right_new, right_tmp
+            right_new, right_tmp = find_next_range(right, right_cur, right_val)
+        elseif isless(left_val, right_val)
+            left_cur, left_val = left_new, left_tmp
+            left_new, left_tmp = find_next_range(left, left_cur, left_val)
+        else
+            right_cur, right_val = right_new, right_tmp
+            right_new, right_tmp = find_next_range(right, right_cur, right_val)
+        end
+    end
+
+    return left_ixs, right_ixs
+end
+
+# optimistically assume that shorter table does not have duplicates in on column
+function _innerjoin_unsorted(left::AbstractArray, right::AbstractArray{T}) where {T}
+    dict = Dict{T, Int}()
+
+    for (idx_r, val_r) in enumerate(right)
+        dict_index = Base.ht_keyindex2!(dict, val_r)
+        dict_index > 0 && return _innerjoin_dup(left, right, dict, idx_r)
+        Base._setindex!(dict, idx_r, val_r, -dict_index)
```

Comment:
We cannot, because I have to switch to another algorithm (`_innerjoin_dup`) if there is a duplicate.
I could to `haskey` check, but this would slow things down, as in case there is no duplicate we would have to repeat the work we do.

The same is below - I could use `haskey` and `getindex`, but it would duplicate the work.

Reply:
I have added comments to the code.